packages/vscode-extension/src/devices/preview.ts

```diff
@@ -194,7 +204,34 @@ export class Preview implements Disposable {
     return this.saveMultimediaWithID(MultimediaType.Screenshot, "screenshot");
   }
 
-  public sendTouches(touches: Array<TouchPoint>, type: "Up" | "Move" | "Down") {
+  public sendTouches(
+    touches: Array<TouchPoint>,
+    type: "Up" | "Move" | "Down",
+    rotation: DeviceRotation
+  ) {
+    // transform touch coordinates to account for different device orientations before
+    // translation and sending them to the simulator-server.
+    touches.forEach((touch, i, array) => {
+      const { xRatio: x, yRatio: y } = touch;
+      switch (rotation) {
+        // 90° anticlockwise map (x,y) to (1-y, x)
+        case DeviceRotation.LandscapeLeft:
+          array[i] = { xRatio: 1 - y, yRatio: x };
+          break;
+        case DeviceRotation.LandscapeRight:
+          // 90° clockwise map (x,y) to (y, 1-x)
+          array[i] = { xRatio: y, yRatio: 1 - x };
+          break;
+        case DeviceRotation.PortraitUpsideDown:
+          // 180° map (x,y) to (1-x, 1-y)
+          array[i] = { xRatio: 1 - x, yRatio: 1 - y };
+          break;
+        default:
+          // Portrait mode: no transformation needed
+          break;
+      }
+    });
```

Comment:
```suggestion
    touches = touches.map((touch) => {
      const { xRatio: x, yRatio: y } = touch;
      switch (rotation) {
        // 90° anticlockwise map (x,y) to (1-y, x)
        case DeviceRotation.LandscapeLeft:
          return { xRatio: 1 - y, yRatio: x };
        case DeviceRotation.LandscapeRight:
          // 90° clockwise map (x,y) to (y, 1-x)
          return { xRatio: y, yRatio: 1 - x };
        case DeviceRotation.PortraitUpsideDown:
          // 180° map (x,y) to (1-x, 1-y)
          return { xRatio: 1 - x, yRatio: 1 - y };
        case DeviceRotation.Portrait:
          // Portrait mode: no transformation needed
          return touch;
      }
    });
```
Use `Array.map` in cases where you want to calculate some function for each element of the array and collect the results.
Use `Array.forEach` when you want to call some function where you only care about some side effect of the function, rather than performing some pure calculation.

Reply:
If you go this way, it might be better creating a new variable `const transformedTouches` for the results of the map instead of reassigning to `touches`.

Reply:
Changed in 5db3f3ac, 5c54d7e5

```ts
public sendTouches(
    touches: Array<TouchPoint>,
    type: "Up" | "Move" | "Down",
    rotation: DeviceRotation
  ) {
    // transform touch coordinates to account for different device orientations before
    // translation and sending them to the simulator-server.
    const transformedTouches = touches.map((touch, i) => {
      const { xRatio: x, yRatio: y } = touch;
      switch (rotation) {
        // 90° anticlockwise map (x,y) to (1-y, x)
        case DeviceRotation.LandscapeLeft:
          return { xRatio: 1 - y, yRatio: x };
        case DeviceRotation.LandscapeRight:
          // 90° clockwise map (x,y) to (y, 1-x)
          return { xRatio: y, yRatio: 1 - x };
        case DeviceRotation.PortraitUpsideDown:
          // 180° map (x,y) to (1-x, 1-y)
          return { xRatio: 1 - x, yRatio: 1 - y };
        default:
          // Portrait mode: no transformation needed
          return touch;
      }
    });

    const touchesCoords = transformedTouches.map((pt) => `${pt.xRatio},${pt.yRatio}`).join(" ");
    this.subprocess?.stdin?.write(`touch ${type} ${touchesCoords}\n`);
  }
  ```